### PR TITLE
feat(net): support eth/68

### LINKED ISSUE #1326: Support eth/68
### Describe the feature

We're missing the eth/68 protocol implementation.
1. Spec: https://eips.ethereum.org/EIPS/eip-5793
2. Geth PR https://github.com/ethereum/go-ethereum/pull/25980

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by jinsankim**
Closes #1326

### Motivation
* Implement the eth/68 protocol

### Reference
* Spec: https://eips.ethereum.org/EIPS/eip-5793
* Geth PR: https://github.com/ethereum/go-ethereum/pull/25980

### CODE REVIEW

File: crates/net/eth-wire/src/types/message.rs
Comment (mattsse) [inline crates/net/eth-wire/src/types/message.rs:158]: this is a bit cursed -.- that the variant now is an enum

File: crates/net/network/src/session/active.rs
Comment (mattsse) [inline crates/net/network/src/session/active.rs:185]: @Rjected perhaps we should apply this check in the stream directly? so the stream only yields well-formatted messages

File: crates/net/eth-wire/src/types/message.rs
Comment (mattsse) [inline crates/net/eth-wire/src/types/message.rs:25]: so now decode_message depends on the version? because `NewPooledTransactionHashes` format differs?

...
I feel like mixing this together with decode is a bit weird. for example decoding GetNodeData now fails with an rlp error even though it can still be valid rlp.

I really would like to do this in two steps:
decode then validate against version.
wdyt @Rjected

File: crates/net/eth-wire/src/hello.rs
Comment (mattsse) [inline crates/net/eth-wire/src/hello.rs:103]: I don't think we should enable this by default already?

**Review by mattsse — CHANGES_REQUESTED** [review]
thanks for taking this on.
left some comments, but would like to hear @Rjected s opinion on some parts.

File: crates/net/eth-wire/src/hello.rs
Comment (rkrasiuk) [reply crates/net/eth-wire/src/hello.rs:103]: agreed, this is a future upgrade, `67` is still considered the current version

File: crates/net/eth-wire/src/types/message.rs
Comment (Rjected) [reply crates/net/eth-wire/src/types/message.rs:158]: Yeah, I think given we don't have support for multiple capabilities, which also has the need for new message types and encodings, it's up in the air how we should do this properly. But yeah, I agree this doesn't seem like the best way to do it

File: crates/net/network/src/session/active.rs
Comment (Rjected) [reply crates/net/network/src/session/active.rs:185]: yes, we should be doing these checks in the stream, and malformed messages should be considered invalid

File: crates/net/eth-wire/src/types/message.rs
Comment (Rjected) [reply crates/net/eth-wire/src/types/message.rs:25]: yes, decoding should not worry about version details, I think this is another consequence of not supporting multiple capabilities. We don't know what to decode into if we don't know the eth wire protocol version being used. It's ambiguous whether or not we should attempt to decode a `eth/68` pooled hashes message, or a `eth/67` pooled hashes message when we receive a `NewPooledTransactionHashes` message ID.

**Review by Rjected — COMMENTED** [review]
I'm really not sure that we should implement this without first figuring out the best way to support multiple capabilities. `eth/66` and `eth/67` are a convenient exception to the rule that most versions have entirely separate message encodings, that are ambiguous unless you know the version. New capabilities also have this problem, so if we push to support `eth/68` without first figuring out how to support other capabilities, I think we might bury a lot of version-specific code / tech debt that should not be there long-term

File: crates/net/eth-wire/src/hello.rs
Comment (jinsankim) [reply crates/net/eth-wire/src/hello.rs:103]: agreed. I thought it's safe because it's still including `66`, `67`, but I was too naive and aggressive. Definitely, I agree not to enable it yet and need to verify it.

Comment (jinsankim) [reply crates/net/eth-wire/src/hello.rs]: follow it up with 651fc4ddee1565cfb935695b8ac225e3b8321edb

File: crates/net/eth-wire/src/ethstream.rs
Comment (jinsankim) [reply crates/net/eth-wire/src/ethstream.rs]: follow it up with ca74056dc177d2724d37ebe9e3b31e744b02b3a2

File: crates/net/eth-wire/src/types/message.rs
Comment (jinsankim) [reply crates/net/eth-wire/src/types/message.rs:161]: I agree it's too ugly and immature. 

I adopted the convention from `geth`. (They uses the name like `NewPooledTransactionHashesPacket66` and `NewPooledTransactionHashesPacket68`.) And I tried to minimize changes. This code is the result of this approach.

I'll think more what's better way.

File: crates/net/eth-wire/src/types/message.rs
Comment (jinsankim) [reply crates/net/eth-wire/src/types/message.rs]: For support multiple capabilities, we must have a branch at some level. It could be just `if`, `match`, `traits`, `module` or whatever. It's needed but I agree that it's not the best way.

**Comment by jinsankim** [comment]
@mattsse @Rjected 

Thank you for comments! I'm much learning from your comments. And now I understand it could cause tech debt and we need to figure out what's the best way to support multiple protocol. 

Please feel free to close it if this PR is not proper way. I just picked it up because the issue is marked as `good-first-issue`. It's enough for me to work it on and learn from it. Whether it could be merged or not, I'd like to revise it in order to learn more. I open to close it any time without any comment. :)

File: crates/net/eth-wire/src/capability.rs
Comment (mattsse) [inline crates/net/eth-wire/src/capability.rs:125]: since we expect that client will update over time we can change the order here: 68 || 67 || 66

File: crates/net/eth-wire/src/types/message.rs
Comment (mattsse) [reply crates/net/eth-wire/src/types/message.rs:161]: I actually like that we do version-specific encoding on the ethstream level.

so, after some thinking this change seems fine to me, because we already have outdated variants like GetNodeData etc.
and because the stream does not emit them if they don't match the version this is ok.

Comment (mattsse) [reply crates/net/eth-wire/src/types/message.rs:161]: I'd like to rename the existing variant to `NewPooledTransactionHashes66` then

File: crates/net/eth-wire/src/types/message.rs
Comment (mattsse) [inline crates/net/eth-wire/src/types/message.rs:161]: also need to update struct docs for EthMessage.

**Review by mattsse — CHANGES_REQUESTED** [review]
alright, we actually need this urgently because this is now a blocker for hive tests.

This looks pretty good already, I guess biggest change is: move tx hash 68 validation into stream itself.

and separate decoding from version but the stream should validate the decoded message against the version, right @Rjected ?

**Comment by Rjected** [comment]
> and separate decoding from version but the stream should validate the decoded message against the version, right @Rjected ?

I guess it might be difficult to decode without using the version, because otherwise we don't know if we should attempt a `NewPooledTransactionHashes68::decode`, or a `NewPooledTransactionHashes::decode`

File: crates/net/eth-wire/src/types/message.rs
Comment (jinsankim) [reply crates/net/eth-wire/src/types/message.rs:161]: > I actually like that we do version-specific decoding on the ethstream level.

agreed and I had tried this way too. I tried to make ethstream as generic depending on eth version. I thought this way is clean but I couldn't implement it with Rust.

Here is my trial (but failed): https://github.com/jinsankim/reth/commit/675d7ec25f6b15f9977f7d59c5c59381bee853a9

**Comment by jinsankim** [comment]
merged `upstream/main` to resolve conflicts.

File: crates/net/eth-wire/src/capability.rs
Comment (jinsankim) [reply crates/net/eth-wire/src/capability.rs:125]: will do

File: crates/net/eth-wire/src/types/message.rs
Comment (jinsankim) [reply crates/net/eth-wire/src/types/message.rs:161]: > I'd like to rename the existing variant to NewPooledTransactionHashes66 then

That's my first implementation but I've reverted it. I'll think again and share why I reverted it. Or will follow up the idea.

**Comment by jinsankim** [comment]
I need to investigate why the test is failed because it seems not to be related with my change. Please note that the work might be delayed because it's weekend in my timezone.

File: crates/net/eth-wire/src/capability.rs
Comment (jinsankim) [reply crates/net/eth-wire/src/capability.rs]: follow it up with b7b4b147cb2d222575d16c35d6dcfb36d16b335b. nit but please note that I also changed the eth versions order as desc in `HelloMessageBuilder`.

File: crates/net/eth-wire/src/types/message.rs
Comment (jinsankim) [reply crates/net/eth-wire/src/types/message.rs:161]: I'm learning for `ActiveSession`, `PeerMessage::PooledTransactions` and so on. The PR might have a bug.

File: crates/net/eth-wire/src/types/message.rs
Comment (jinsankim) [reply crates/net/eth-wire/src/types/message.rs:167]: follow it up with 011eb73e763d5ba6e7ad2f0b3b4cb22d9f58e807. Please suggest if docs is insufficient because I'm not a fluent English speaker.

File: crates/net/eth-wire/src/types/message.rs
Comment (jinsankim) [reply crates/net/eth-wire/src/types/message.rs:167]: After learning code more, I could remember why I choose to use `NewPooledTransactionHashes` than `NewPooledTransactionHashes66`. It's because `NewPooledTransactionHashes` is used internally as well for `PeerMessage::PooledTransactions`, `NetworkHandleMessage::SendPooledTransactionHashes` and `NetworkTransactionEvent::IncomingPooledTransactionHashes`. 

Now I've realized that we should care of these as well. maybe need more `66`, `68` convention and branches. T_T I'm working on it.

Comment (jinsankim) [reply crates/net/eth-wire/src/types/message.rs]: > I'd like to rename the existing variant to NewPooledTransactionHashes66 then

follow it up with a94b1da162f48db190c321fa8068b457bc439ce2. I've changed this PR into Draft because the new change is little big. I need to add more tests and review myself first.

Comment (jinsankim) [reply crates/net/eth-wire/src/types/message.rs]: ready for review.

**Comment by jinsankim** [comment]
I've worked more especially for sending `NewPooledTransactionHashes`. PTAL carefully and thanks in advance for any comments and advice.

File: crates/net/eth-wire/src/types/message.rs
Comment (gakonst) [inline crates/net/eth-wire/src/types/message.rs:26]: yeah this won't work if there's other capabilities in the future, we need to separate the validation. i guess it's fine for now.

**Review by gakonst — CHANGES_REQUESTED** [review]

**Comment by gakonst** [comment]
Nice work! Some nits from my side. @mattsse wdyt

File: crates/net/eth-wire/src/capability.rs
Comment (jinsankim) [reply crates/net/eth-wire/src/capability.rs:73]: I think it's better to open another issue for it.

File: crates/transaction-pool/src/lib.rs
Comment (jinsankim) [reply crates/transaction-pool/src/lib.rs:228]: the dependency of `struct`s is a little bit complecated. I'll try to resolve it.

File: crates/net/network/src/session/active.rs
Comment (jinsankim) [reply crates/net/network/src/session/active.rs]: follow it up with d5da4f66b5b213d9d7317044aa7edff1225cc07c.

File: crates/transaction-pool/src/lib.rs
Comment (jinsankim) [reply crates/transaction-pool/src/lib.rs]: follow it up with e1d007b297479b68db6c6c3078a51edda2b7f799.

**Review by gakonst — CHANGES_REQUESTED** [review]
Good work! I think final things missing might be some tests? Defer to @mattsse otherwise. I guess passing Hive will matter there, but worth configuring 2 clients with eth/68 and seeing that it works? Or one with 67 and one with 68?

File: crates/net/eth-wire/src/errors/eth.rs
Comment (mattsse) [inline crates/net/eth-wire/src/errors/eth.rs:19]: I'd like to use named fields here, otherwise it's harder to understand what the different usizes are

File: crates/net/eth-wire/src/types/broadcast.rs
Comment (mattsse) [inline crates/net/eth-wire/src/types/broadcast.rs:134]: not a fan of this conversion.

why would we need this anyway?
I'd like to remove this.

File: crates/net/eth-wire/src/types/message.rs
Comment (mattsse) [reply crates/net/eth-wire/src/types/message.rs:76]: I think since this is no longer rlp only we need to change the error type we return as well and add a dedicated error variant for when this happens.

File: crates/net/network/src/message.rs
Comment (mattsse) [inline crates/net/network/src/message.rs:44]: why do we need this, why do we need optional fields here?

File: crates/net/network/src/message.rs
Comment (mattsse) [inline crates/net/network/src/message.rs:69]: From impls for tuples are not great so would abandon them entirely.

but not clear why we need this type at all.

**Review by mattsse — CHANGES_REQUESTED** [review]
only a few suggestion for the eth-wire changes.

I have some questions regarding the network changes, mainly why we need this type with optional fields.

can we extract all of these and submit separately, so we can merge the eth68 changes asap?

**Comment by jinsankim** [comment]
@gakonst 
> worth configuring 2 clients with eth/68 and seeing that it works? Or one with 67 and one with 68?

I've tested and succeeded it with enabling only eth/68 and connecting to `mainnet` but I think it's conducted only early stage. Because the PR is too old during reviewing, I'll test it again.

**Comment by jinsankim** [comment]
@mattsse 
> I have some questions regarding the network changes, mainly why we need this type with optional fields.

Thanks for your comment! To my understanding, maybe you asked why does `NewPooledTransactionHashes` have optional `types` and `sizes`, right? If I understand correctly, because it's used for `PeerMessage::PooledTransactions`.

https://github.com/paradigmxyz/reth/blob/058349ecf86a41dd71b4916dc28cefa56eb8fb7e/crates/net/network/src/message.rs#L40-L59

`PeerMessage::PooledTransactions` is bi-directional eth-message variants so it could represent `NewPooledTransactionHashes66` and `NewPooledTransactionHashes68`. As you know, `NewPooledTransactionHashes66` has only `hashes` but `NewPooledTransactionHashes68` has all `types`, `sizes`, and `hashes`.

My first approach was decomposing `PooledTransactions` into each `PooledTransactions66` and `PooledTransactions68`. But after some trial, I've realized that it's better to use optional field to simplify handlers.

I wonder whether it's acceptable or not. Always welcome for any advice.

**Comment by mattsse** [comment]
I see, I'd like to add another variant for 68 here as well, but let's do that separately, we can leave incoming 68 unhandled in this PR

File: crates/net/eth-wire/src/errors/eth.rs
Comment (jinsankim) [reply crates/net/eth-wire/src/errors/eth.rs]: follow it up with 231f46085521ada2ea6697bdd8f83baf971faa35.

File: crates/net/network/src/message.rs
Comment (jinsankim) [reply crates/net/network/src/message.rs]: follow it up with acb0d81f90e16bdd4c9070208470d7ff0b46b3c5.

File: crates/net/eth-wire/src/types/broadcast.rs
Comment (jinsankim) [reply crates/net/eth-wire/src/types/broadcast.rs:134]: Please give me an advice. T_T

* It's used for converting from `NewPooledTransactionHashes` into `NewPooledTransactionHashes68`.
* `NewPooledTransactionHashes` is defined in `network` so we could not define `TryFrom<NewPooledTransactionHashes> for NewPooledTransactionHashes68`.
* The only alternative idea I have is that `pub fn new(...) -> Result<Self, Error>`.

HDYT?

File: crates/net/eth-wire/src/types/message.rs
Comment (jinsankim) [reply crates/net/eth-wire/src/types/message.rs]: I think at this moment `EthStreamError` is the proper one.

follow it up with c70e2d25d43eb9f95e1ca3e2bfe5d4bf5dcd4bcf.

File: crates/net/eth-wire/src/types/broadcast.rs
Comment (mattsse) [reply crates/net/eth-wire/src/types/broadcast.rs:134]: do we need this in this PR?

I still don't understand why we would need this.

in any case, a TryFrom a tuple with 3 items is not great.

**Comment by mattsse** [comment]
@jinsankim can we please do the bare minimum for eth68 here and do the new message types in a separate PR?

**Comment by jinsankim** [comment]
@mattsse 
> can we please do the bare minimum for eth68 here and do the new message type in a separate PR?
little bit confusing. I need more guide about what's the bare minimum for eth68. I agree whatever we want for this PR, close, modify, revert only some part or whatever. But I can't catch what should I do for bare minimum.

File: crates/net/network/src/message.rs
Comment (mattsse) [inline crates/net/network/src/message.rs:44]: please remove this type and instead move to separate PR

File: crates/net/eth-wire/src/types/broadcast.rs
Comment (mattsse) [inline crates/net/eth-wire/src/types/broadcast.rs:134]: please remove this trait impl entirely.

**Comment by jinsankim** [comment]
@mattsse 

What I've caught your direction from 2 todos,

* minimize changes 
* only focusing incoming (even though it's adhoc)
* remove outgoing

Am I understanding correctly?

**Comment by mattsse** [comment]
yeh, I want to merge this with just 68 support on the eth wire level to unblock hive tests.
and followup with the actual 68 tx handling (in network impl) separately

**Comment by jinsankim** [comment]
I have no knowledge for hive tests. But I've just concerned that if we send `NewPooledTransactionHashes66` though eth/68, maybe tests will be failed.

I agree to merge PR step by step!

File: crates/net/network/src/message.rs
Comment (jinsankim) [reply crates/net/network/src/message.rs]: follow it up with f53e2a7f9a19eb19ffe05dc86a74b84867a3a8c2.

File: crates/net/eth-wire/src/types/broadcast.rs
Comment (jinsankim) [reply crates/net/eth-wire/src/types/broadcast.rs]: follow it up with f53e2a7f9a19eb19ffe05dc86a74b84867a3a8c2.

**Comment by jinsankim** [comment]
Now, in this PR, it could handle incoming eth/68 message but couldn't handle outgoing eth/68 message. I left a TODO comments for it.

**Review by gakonst — APPROVED** [review]
@mattsse calling this good, the logic for parsing msgs seems fine.

main 2 things are:
1. We do `try_emit_broadcast` w/o types/sizes to `NewPooledTransactionHashes68` msgs, is this correct? Or will we get penalized?
2. The TODO at active.rs##254

**Comment by jinsankim** [comment]
@gakonst 

> We do try_emit_broadcast w/o types/sizes to NewPooledTransactionHashes68 msgs, is this correct? Or will we get penalized?

I think the incoming handling will be working well. `geth` also uses types/sizes only to check validation. Internally we need only hashes at least now.

> The TODO at active.rs#https://github.com/paradigmxyz/reth/issues/254

outgoing is removed from this PR because of unacceptable message structure. I hope it'll be followed up with #254 and also improve architecture for multi protocol in the future.

**Comment by jinsankim** [comment]
> Coverage   76.19%   47.27%   -28.93% 

I think it's too weird. :(

**Pull request merged**
